jupyter_server_fileid/extension.py

```diff
@@ -12,17 +17,26 @@ class FileIdExtension(ExtensionApp):
     name = "jupyter_server_fileid"
 
     file_id_manager_class = Type(
-        klass=FileIdManager,
-        help="File ID manager instance to use. Defaults to FileIdManager.",
+        klass=AbstractFileIdManager,
+        help="""File ID manager instance to use.
+
+        Defaults to:
+        - LocalFileIdManager if contents manager is a FileContentsManager,
+        - ArbitraryFileIdManager otherwise.
+        """,
         config=True,
     )
 
-    @default("file_id_manager")
-    def _file_id_manager_default(self):
-        self.log.debug("No File ID manager configured. Defaulting to FileIdManager.")
-        return FileIdManager
-
     def initialize_settings(self):
+        if self.file_id_manager_class == AbstractFileIdManager:
+            if isinstance(self.settings["contents_manager"], FileContentsManager):
+                self.log.debug(
+                    "Contents manager is a FileContentsManager, defaulting to LocalFileIdManager."
+                )
+                self.file_id_manager_class = LocalFileIdManager
+            else:
+                self.file_id_manager_class = ArbitraryFileIdManager
+
```

Comment:
This is close to what we want, but not quite. This always sets `file_id_manager_class` even if the user specifies a different one via Jupyter config. We only want to dynamically set the **default** class, not the class itself.

```python
    @default(file_id_manager_class)
    def _default_file_id_manager_class():
      return LocalFileIdManager if isinstance(self.settings["contents_manager"], FileContentsManager) else ArbitraryFileIdManager
```

Reply:
Hm, it looks like this doesn't work, as `self.settings` is not yet initialized when this class is constructed.